In [1]:
import os
from selenium import webdriver
from selenium.webdriver import Chrome
from selenium.webdriver.common.keys import Keys
import time

## Crawling function

In [2]:
def crawling(place):
    for i in place:
        #browser2로 각 카페 리뷰 사이트에 접속하기
        url = i.get_attribute('href')
        time.sleep(1)
        
        browser2.get(url)
        time.sleep(1)
        
        go = i.text[0]
        
        #카페 상세사이트의 1페이지 리뷰와 별점 수집하기 - 평가 개수가 0이 아닐 때평가 개수가 0이 아닐 때
        if int(go) != 0:
        
            #별점이 없는 '카카오플레이스 평가하기'의 수집은 제외하기 위해서 카카오맵 리뷰만 로딩
            try:
                browser2.find_element_by_css_selector('#mArticle > div.cont_evaluation > div.evaluation_sorting > a').send_keys(Keys.ENTER)
                browser2.find_element_by_css_selector('#mArticle > div.cont_evaluation > div.evaluation_sorting.open_sorting > ul > li:nth-child(2) > a').send_keys(Keys.ENTER)
                time.sleep(1)
            except:
                continue


            #카페 이름 수집하기
            ready_name = browser2.find_element_by_css_selector('div.inner_place h2.tit_location').text
            time.sleep(3)

  
            [review.append(r.text) for r in browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')]
            [score.append(s.text) for s in browser2.find_elements_by_css_selector('div.grade_star.size_s em.num_rate')]
            [name.append(ready_name) for n in range(len(browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')))]




            ##카페 상세사이트의 2~6페이지 리뷰와 별점 수집하기
            while True:
                idx=3
                try:
                    #다음버튼 전까지의 숫자페이지 리뷰와 별점 수집하기
                    page_num = len(browser2.find_elements_by_class_name('link_page'))
                    for i in range(page_num-1):
                        browser2.find_element_by_css_selector('#mArticle > div.cont_evaluation > div.evaluation_review > div > a:nth-child(' + str(idx) +')').send_keys(Keys.ENTER)
                        #2페이지가 (#mArticle > div.cont_evaluation > div.evaluation_review > div > a:nth-child(3))
                        time.sleep(1)

                        [review.append(r.text) for r in browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')]
                        [score.append(s.text) for s in browser2.find_elements_by_css_selector('div.grade_star.size_s em.num_rate')]
                        [name.append(ready_name) for n in range(len(browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')))]

                        idx = idx +1

                    #다음버튼을 누른 후 다음페이지 리뷰와 별점 수집하기
                    browser2.find_element_by_link_text('다음').send_keys(Keys.ENTER)
                    time.sleep(1)

                    [review.append(r.text) for r in browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')]
                    [score.append(s.text) for s in browser2.find_elements_by_css_selector('div.grade_star.size_s em.num_rate')]
                    [name.append(ready_name) for n in range(len(browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')))]

                except:
                    break

            #6~끝 페이지 리뷰와 별점 수집하기
            while True:
                idx=4
                try:
                    #다음버튼 전까지의 숫자페이지 리뷰와 별점 수집하기
                    page_num = len(browser2.find_elements_by_class_name('link_page'))
                    for i in range(page_num-1):
                        browser2.find_element_by_css_selector('#mArticle > div.cont_evaluation > div.evaluation_review > div > a:nth-child(' + str(idx) +')').send_keys(Keys.ENTER)
                        time.sleep(1)

                        [review.append(r.text) for r in browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')]
                        [score.append(s.text) for s in browser2.find_elements_by_css_selector('div.grade_star.size_s em.num_rate')]
                        [name.append(ready_name) for n in range(len(browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')))]

                        idx = idx +1

                    #다음버튼을 누른 후 다음페이지 리뷰와 별점 수집하기
                    browser2.find_element_by_link_text('다음').send_keys(Keys.ENTER)
                    time.sleep(1)

                    [review.append(r.text) for r in browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')]
                    [score.append(s.text) for s in browser2.find_elements_by_css_selector('div.grade_star.size_s em.num_rate')]
                    [name.append(ready_name) for n in range(len(browser2.find_elements_by_css_selector('div.comment_info p.txt_comment')))]
                except:
                    print(browser2.find_element_by_css_selector('div.inner_place h2.tit_location').text, "데이터 수집 완료.") 
                    break
        else:
            continue

## Completed Crawling code
- '강릉 카페'

In [5]:
#셀레늄 연결
browser = Chrome()
browser2 = Chrome()
time.sleep(1)

#browser1에 카카오맵 연결하기
browser.get('https://map.kakao.com/')
time.sleep(1)

#검색창에 검색어 입력
search_area = browser.find_element_by_css_selector('#search\.keyword\.query')
search_area.send_keys('강릉 카페')
time.sleep(1)

#검색하기 클릭
browser.find_element_by_css_selector('.go.ico_search.btn_search').send_keys(Keys.ENTER)
time.sleep(1)

#리스트 생성
name = []
review = []
score = []

#browser를 통한 1페이지 크롤링
cafes = browser.find_elements_by_css_selector('a.numberofscore')
time.sleep(1)
crawling(cafes)

#우선 1페이지의 더보기를 클릭
browser.find_element_by_css_selector('#info\.search\.place\.more').send_keys(Keys.ENTER)
time.sleep(1)

#2페이지~마지막페이지
while True:
    idx=2    
    try:
        #다음버튼 전까지의 페이지 크롤링 진행 후 페이지넘기기
        page_num = len(browser.find_elements_by_css_selector('div.pages'))
        for i in range(page_num-3):
            browser.find_element_by_css_selector('#info\.search\.page\.no'+ str(idx) +'').send_keys(Keys.ENTER)
            time.sleep(1)
            cafes = browser.find_elements_by_css_selector('a.numberofscore')
            time.sleep(1)
            crawling(cafes)
            idx = idx +1

        #다음버튼을 누른 후 해당 페이지 크롤링 진행 후 페이지넘기기
        browser.find_element_by_css_selector('#info\.search\.page\.next').send_keys(Keys.ENTER)
        time.sleep(1)
        cafes = browser.find_elements_by_css_selector('a.numberofscore')
        time.sleep(1)
        crawling(cafes)
    except:
        print('크롤링 끝!')
        break

지베이스 데이터 수집 완료.
카페오로라 데이터 수집 완료.
바빈스커피 영진점 데이터 수집 완료.
테라로사 커피공장 데이터 수집 완료.
곳 데이터 수집 완료.
테라로사 경포호수점 데이터 수집 완료.
보헤미안박이추커피공장 데이터 수집 완료.
돌체테리아 데이터 수집 완료.
테라로사 사천점 데이터 수집 완료.
테라로사 경포대점 데이터 수집 완료.
퍼베이드 데이터 수집 완료.
오월 데이터 수집 완료.
보사노바 강릉본점 데이터 수집 완료.
커피바다 데이터 수집 완료.
테라로사 임당점 데이터 수집 완료.
카페 뤼미에르 데이터 수집 완료.
산토리니 강릉안목본점 데이터 수집 완료.
보헤미안 영진 데이터 수집 완료.
카페기와 데이터 수집 완료.
키크러스커피 데이터 수집 완료.
봉봉방앗간 데이터 수집 완료.
카페툇마루 데이터 수집 완료.
강릉커피빵 본점 데이터 수집 완료.
346커피스토리 데이터 수집 완료.
비사이드그라운드 데이터 수집 완료.
엘빈 데이터 수집 완료.
롱블랙 데이터 수집 완료.
칠커피 데이터 수집 완료.
스타벅스 강릉강문해변점 데이터 수집 완료.
미르마르 데이터 수집 완료.
애시당초 데이터 수집 완료.
보헤미안 데이터 수집 완료.
갈매기다방 데이터 수집 완료.
스타벅스 강릉안목항 데이터 수집 완료.
초당커피정미소 데이터 수집 완료.
쉘리스커피 데이터 수집 완료.
월성회관 데이터 수집 완료.
106 데이터 수집 완료.
100년임당방앗간 데이터 수집 완료.
그다이마잇 데이터 수집 완료.
쎄라비 데이터 수집 완료.
롱브레드 안목점 데이터 수집 완료.
허씨커피 데이터 수집 완료.
러브레터 카페앤레스토랑 데이터 수집 완료.
에이엠브레드 데이터 수집 완료.
카페브라질 영진본점 데이터 수집 완료.
토부 데이터 수집 완료.
도깨비시장 데이터 수집 완료.
스타벅스 강릉교동광장점 데이터 수집 완료.
커피아메리카 데이터 수집 완료.
노브일리 데이터 수집 완료.
로지커피 데이터 수집 완료.
오리까페 데이터 수집 완료.
명주배롱 데이터 수집 완료.
아트리스테이블 데이터 수집 완료.
커피

In [11]:
len(name), len(review), len(score)

(3245, 3245, 3245)

## 크롤링 결과
- 끝까지 크롤링됨.
- 3245개의 데이터 수집
- 저장

In [12]:
import pandas as pd

In [13]:
gangneung_cafe = pd.DataFrame({'name': name, 'review': review, 'score': score})
gangneung_cafe

name                                             review score
0            지베이스  대단한 컬렉션입니다. 건프라 뿐 아니라 피규어와 서적, 각종 한정판 굿즈 까지......     5
1            지베이스  신기했어요! 볼거리도 많고!!! 저희 같은 마니아층에게는 천국같은... 너무 좋았어...     5
2           카페오로라  카페오로라 옥상 장난아닙니다, 인생샷 찍었어요 여기서ㅜㅜ 오로라라떼는 필수고 , 오...     5
3           카페오로라  친절하고 카페가 깨끗하고 좋았어요 커피맛은 무난하게 맛있었고 특히 라떼가 고소하니 ...     4
4           카페오로라  카페가 깔끔하고 이뻐요!!! 커피맛은 개취라지만 저한테는 딱 좋은~~ 옥상에 포토존...     5
...           ...                                                ...   ...
3240       카페미소티스                       풍경이 너무 맘에 드네요 ~ 미리 예약신청해야겠어요     5
3241   토프레소 강릉내곡점                                                좋아요     5
3242       표씨네 카페                            예쁘고 맛있어요. 가격이 저렴해서 좋아요.     5
3243  콩볶는커피집 알레그로     조용하고 좋은 카페입니다. 고소한 커피볶는 냄새가 좋았습니다. 책 잘 읽다 갑니다!     5
3244        카페라우머                                               최고예요     5

[3245 rows x 3 columns]

In [14]:
gangneung_cafe.to_csv('gangneung_cafe.csv')